### DISCUSSION TITLE
Phone audio as source

### DISCUSSION THREAD

**Discussion by Dejv311** [music-assistant/support#384] [Feature requests and ideas]
I would really appreciate being able to connect my phone/pc to home assistant like a source (not over spotify/miracast, but like a bluetooth speaker).

The idea is, that there is no easy way to stream local phone music/pc game sounds/smart tv audio to HA, thus all the products need a separate playback system.

Please vote, if this is something you would like to see too...

**Comment by erkr** [comment]
Thanks for your request

**Comment by Torrax** [comment]
I really like this idea. Would it be possible to implement a music provider that sources from a local soundcard on the running system? I have an external bluetooth receiver with a button so people can quickly connect. It would be awesome to stream that audio though a sound card to the music assistant.

This would also allow us to connect things like projectors to your outdoor speakers, turn your home speakers into a PA system for events. Play from a record player to different zones all around your house. I think there are a bunch of great opportunities here. I'm going to try to take a shot at it myself but might need to pick some brains.

**Reply by OzGav** [reply]
Saw your Discord post as well. There has also been related discussion here https://github.com/orgs/music-assistant/discussions/1169#discussioncomment-5793412

**Comment by arretx** [comment]
This can be done, but it requires a few things.  You can use two of the Dante AVIO adapters to accomplish this goal.  They have a Bluetooth receiver that you connect your mobile device to, and then it connects to your network switch.  Then, you use a Dante Analog output to send that network based audio to an analog amplifier, like a whole house audio amplifier.  Use Music Assistant to select the media player (for example, the Monoprice 6-channel amp creates 6 media players) and then adjust the source for that media-player to be the analog input that the Dante AVIO analog is connected to.  Voila!  You have networked audio.  And, with Dante Virtual Sound Card, any of your computers (hard-wired ethernet) can transmit audio to the Dante network which the AVIO can convert to analog.

**Comment by OzGav** [comment]
The docs describe a way to stream to MA https://music-assistant.io/faq/stream-to/

**Comment by Hedda** [comment]
FYI, for reference, I have previously used the open-source ”Vinyl Cast” app on an Android phone/tablet with a USB Audio Device adapter (Behringer UCA202) for casting analog input from a vinyl record player / LP turntable to Chromecast Audio dongle using Google Cast but it has been very buggy.

https://vinylcast.schober.tech/

https://play.google.com/store/apps/details?id=tech.schober.vinylcast.playstore

https://github.com/aschober/vinyl-cast

https://youtu.be/HBDkxEvCcHQ

PS: The initial release of the Vinyl Cast app also included using Audio ACR to detect the song being played, but unfortunately this feature relied on was discontinued in third-party library/service so the feature was removed.

**Comment by JHarding86** [comment]
This would be a very welcome addition in my household. 

I could see this happening in a one of two ways. 

1. Bluetooth music provider (Preffered method)

Create a bluetooth music provider that allows me to connect my phone / tablet / whatever via Bluetooth to the Music Assistant server. Then route this music to the player provider that I have chosen within Music Assistant. 

2. Emulate a Chromecast / Airplay receiver

Create a music provider that emulates a Chromecast. Then from the device of my choosing I could cast my devices audio to the Music Assistant server which would the route the music to the player provider I have chosen within Music Assistant.

**Reply by N8schatten** [reply]
Yes, the Bluetooth method would also suit me very much

**Reply by djgenesis** [reply]
I am also interested in something like this. Pass audio from my phone to HA either via BT or do a Chromecast emulation.

**Reply by Rafael-Amonal** [reply]
I am more interested in the Chromecast or AirPlay emulator. Since my wife and I share a Bluetooth speaker for the shower, I use it for playing YouTube videos and she uses it for listening to music and connecting and disconnecting our phone is a pain. I'd prefer to transform the Bluetooth speaker into a home assistant media player and cast the audio to it.

**Reply by TheBlueKingLP** [reply]
Chromecast emulation would be hard as each device seems to require a certificate to be recognized as a valid receiver by any official Chromecast source/sender. Someone could create their way of sending audio but capturing that audio from some DRM protected audio source would be tricky/impossible. Thus Bluetooth would still be wanted.

**Reply by OzGav** [reply]
Have a look at the Bluetooth sendspin bridge

**Reply by TheBlueKingLP** [reply]
> Have a look at the Bluetooth sendspin bridge

Just want to confirm, it can send audio from a phone into Music Assistant? i.e. Music Assistant act as a "Bluetooth speaker" by pairing phone to a Bluetooth adapter and the phone will see the adapter as a "speaker". I installed it and can't seems to find that option.

**Reply by OzGav** [reply]
If you want to go that way and have an Android phone then use Ariacast

**Reply by TheBlueKingLP** [reply]
> If you want to go that way and have an Android phone then use Ariacast

Right, it does seems to work with some app but refuses to cast with other apps such as YouTube. I suspect it is due to YouTube blocking capturing. Correct me if I am wrong or if this is somehow just me.

**Reply by OzGav** [reply]
I dont use it so best to ask the maintainer about that. Sorry I cant help.

**Reply by gartensofa** [reply]
Yes, that is documented behavior. 
Read the line for advanced users though:
https://github.com/AriaCast/AriaCast-app#-audio-source-compatibility

**Comment by raphaeldelio** [comment]
By implementing bluetooth as an input source, we can also easily use any output source as a speaker for Alexa. 

This would allow Homepods and Apple TVs to be used as an output source for Alexa.

**Comment by DanielKinsman** [comment]
I managed to get this working on a standalone sendspin server (outside of music assistant) with [a small change to sendspin-cli](https://github.com/Sendspin/sendspin-cli/pull/102) and some pipewire magic.

I don't know if it will play well with music assistant also running at the same time, reading the spec it would seem [multiple servers](https://github.com/Sendspin/spec?tab=readme-ov-file#multiple-servers) aren't handled all that well.

**Comment by Hedda** [comment]
> I managed to get this working on a standalone sendspin server (outside of music assistant) with https://github.com/Sendspin/sendspin-cli/pull/102 and some pipewire magic. I don't know if it will play well with music assistant also running at the same time, reading the spec it would seem [multiple servers](https://github.com/Sendspin/spec?tab=readme-ov-file#multiple-servers) aren't handled all that well.

Sendspin specification developers have previously written that they have no plans to add support for multiple servers and that they instead plan to add a new source/sender role to Sendspin client-side, so you can have a single server and yet allow each client to also act a source too.

FYI; someone submitted a draft pull request which would add support for such a source/sender role to Sendspin-clients here:

* https://github.com/Sendspin/spec/pull/52

For reference; there is also a direcly related discussion here about how support for audio input could be added to the Sendspin protocol spec:

* https://github.com/Sendspin/spec/issues/14

**Comment by OzGav** [comment]
Ariacast will soon appear in the beta which will allow streaming from Android devices

**Reply by N8schatten** [reply]
Great, that sounds interesting. Thanks!

**Comment by sprokos1980** [comment]
Cheers, everyone! I managed to achieve a very similar result using a spare Raspberry Pi, on which I installed piCorePlayer.
In PCP, after enabling Bluetooth from the main page and pairing my phone, I changed its type from Player to Streamer. From the Tweaks tab, I enabled streamer in audio tweaks which provided me with a basic URL.
I then added this URL as a radio station in Music Assistant. Now I can stream my phone’s audio to any speaker in my house, or use Snapcast to stream music (or any content like Spotify, YouTube, etc.) to all my speakers simultaneously.
No coding or technical skills needed—just a spare RPi. The audio quality is decent (it is Bluetooth, after all); there’s just a latency of about 2–3 seconds.

**Reply by Cytomax55** [reply]
Are you still using this method, have you found any bugs with it or is it working as expected?

**Reply by sprokos1980** [reply]
Everything still works great. I think the easiest option is to use ariacast, you wont even need extra hardware

**Reply by N8schatten** [reply]
Only problem is that there is no chance to send drm protectet audio

**Comment by Cytomax55** [comment]
I would very much like a feature where I can connect my phone via Bluetooth to play things like Pandora and or YouTube music... Are there any updates to this?

**Comment by Hedda** [comment]
FYI, Sendspin specification was just extended to add basic source role support meant for audio input, allows a client to capture local audio (line-in, turntable, Bluetooth receiver, microphone) and stream it to the server, but it is brand new so specs only so far and no implementations, see:

* https://github.com/Sendspin/spec/pull/105

and

  * https://github.com/Sendspin/spec/issues/14

Also, take note and be aware that **ESP32-S31** is a brand new MCU SoC hardware chip from Espressif that was just now made availalable which feature support for Bluetooth Sink, so maybe that could be used as a Bluetooth audio receiver to capture the audio and forward to Sendspin server in Music Assistent  in the future?

* https://www.espressif.com/en/news/ESP32_S31_Release

This new ESP32-S31 can be seen as a RISC-V based successor to the ESP32-S3, and it offers more modern connectivity options and feature 1Gbps Ethernet MAC Ethernet, Wi-Fi 6 (2.4GHz-only), Bluetooth 5.4, and IEEE 802.15.4 (Thread/Zigbee) support.

   * https://www.espressif.com/en/products/socs/esp32-s31

In summery when comparing ESP32-S31 to the ESP32-S3, unlike the XTensa architecture of the ESP32-S3, the new ESP32-S31 features a high-performance dual-core 32-bit RISC-V processor running at maximum of 320 MHz (compared to the the ESP32-S3's 240 MHz). ESP32-S31 also features much higher memory bandwidth with high-speed 250 MHz 8-bit DDR PSRAM (which give it a relativly massive speed boost compared to the standard Octal PSRAM in the ESP32-S3) . The high-performance core in the ESP32-S31 also features a 128-bit wide data path with SIMD (Single Instruction, Multiple Data) and FPU extensions for hardware-level vector acceleration provides the math heavy-lifting to handle complex compute directly on-chip, (like example  driving a large LVGL UI on a larger display). 

These ESP32-S31 dev boards and usable modules are supposed to be availabile or comming soon:

* **ESP32-S31-WROOM-3 module** (ESP32-S31-WROOM-3) - Official wireless radio module from Espressif
* https://documentation.espressif.com/esp32-s31-wroom-3_datasheet_en.pdf

* **Wireless-Tag ESP32-S31 Tiny (ESP32S31-TINY)** - looks to be coming soon but does not look to be available yet.
  * https://www.youtube.com/watch?v=guJj771c1cE
  * https://www.youtube.com/watch?v=RnS8VM__PNY
  * https://www.youtube.com/watch?v=yMXlIkspvYU

* **ESP32-S31-Korvo-1 multimedia development board** - An all-in-one 4.3-inch LCD 800 x 480 resolution display and camera.
* https://docs.espressif.com/projects/esp-dev-kits/en/latest/esp32s31/esp32-s31-korvo-1/index.html

* **ESP32-S31-Function-CoreBoard-1** is a development board for connected AIoT prototyping based on the **ESP32-S31-WROOM-3 module**. It supports Wi-Fi 6, IEEE 802.15.4, Bluetooth 5.4 (LE + BR/EDR), and Gigabit Ethernet; provides an onboard microphone and speaker output; routes out all key GPIOs. Most of the I/O pins on the module are broken out to pin header J2 for easy interfacing. It also features a USB 2.0 Type-A Port and USB Type-C port which can be used for USB pheripials (though the USB Type-C port will require a USB-hub if you also want to power the board via USB).
* https://docs.espressif.com/projects/esp-dev-kits/en/latest/esp32s31/esp32-s31-function-coreboard-1/index.html

References:

* https://www.reddit.com/r/esp32/comments/1tmtswg/esp32s31_is_out_yall/

For detailed specification description and development kit hardware overview check out:

* https://www.cnx-software.com/2026/05/22/esp32-s31-development-boards-iot-smart-audio-and-hmi-applications/

and

* https://linuxgizmos.com/esp32-s31-development-boards-bring-wi-fi-6-audio-camera-and-hmi-features/

**Reply by Cytomax55** [reply]
Ty for the update, so in summary the new spec is in sendspin but we need people to figure out how to implement it, and there is new powerful hardware that can be used to implement it, so positive news all around, ty for updating me
